--- ../net/minecraft/network/packet/Packet3Chat.java
+++ ../net/minecraft/network/packet/Packet3Chat.java
@@ -2,7 +2,8 @@
 
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
-import java.io.IOException;
+
+import java.io.IOException; // CraftBukkit
 
 public class Packet3Chat extends Packet
 {
@@ -23,22 +24,21 @@
     public Packet3Chat(String p_i3301_1_, boolean p_i3301_2_)
     {
         this.field_73477_c = true;
-
-        if (p_i3301_1_.length() > field_73478_a)
-        {
-            p_i3301_1_ = p_i3301_1_.substring(0, field_73478_a);
+        /* CraftBukkit start - handle this later
+        if (s.length() > b) {
+            s = s.substring(0, b);
         }
-
+        // CraftBukkit end */
         this.field_73476_b = p_i3301_1_;
         this.field_73477_c = p_i3301_2_;
     }
 
-    public void func_73267_a(DataInputStream p_73267_1_) throws IOException
+    public void func_73267_a(DataInputStream p_73267_1_) throws IOException   // CraftBukkit
     {
         this.field_73476_b = func_73282_a(p_73267_1_, field_73478_a);
     }
 
-    public void func_73273_a(DataOutputStream p_73273_1_) throws IOException
+    public void func_73273_a(DataOutputStream p_73273_1_) throws IOException   // CraftBukkit
     {
         func_73271_a(this.field_73476_b, p_73273_1_);
     }
