--- ../net/minecraft/inventory/ContainerBeacon.java
+++ ../net/minecraft/inventory/ContainerBeacon.java
@@ -1,9 +1,14 @@
 package net.minecraft.inventory;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryBeacon;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntityBeacon;
+// CraftBukkit end
 
 public class ContainerBeacon extends Container
 {
@@ -12,9 +17,14 @@
     private int field_82865_g;
     private int field_82867_h;
     private int field_82868_i;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    // CraftBukkit end
 
     public ContainerBeacon(InventoryPlayer p_i5076_1_, TileEntityBeacon p_i5076_2_)
     {
+        player = p_i5076_1_; // CraftBukkit
         this.field_82866_e = p_i5076_2_;
         this.func_75146_a(this.field_82864_f = new SlotBeacon(this, p_i5076_2_, 0, 136, 110));
         byte b0 = 36;
@@ -59,6 +69,11 @@
 
     public boolean func_75145_c(EntityPlayer p_75145_1_)
     {
+        if (!this.checkReachable)
+        {
+            return true;    // CraftBukkit
+        }
+
         return this.field_82866_e.func_70300_a(p_75145_1_);
     }
 
@@ -126,4 +141,18 @@
 
         return itemstack;
     }
+
+    // CraftBukkit start
+    public CraftInventoryView getBukkitView()
+    {
+        if (bukkitEntity != null)
+        {
+            return bukkitEntity;
+        }
+
+        CraftInventory inventory = new CraftInventoryBeacon(this.field_82866_e);
+        bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
