--- ../net/minecraft/entity/ai/EntityAITarget.java
+++ ../net/minecraft/entity/ai/EntityAITarget.java
@@ -1,5 +1,7 @@
 package net.minecraft.entity.ai;
 
+import org.bukkit.event.entity.EntityTargetEvent; // CraftBukkit
+import net.minecraft.entity.EntityCreature;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.passive.EntityTameable;
 import net.minecraft.entity.player.EntityPlayer;
@@ -144,6 +146,51 @@
                     }
                 }
 
+                // CraftBukkit start - check all the different target goals for the reason, default to RANDOM_TARGET
+                EntityTargetEvent.TargetReason reason = EntityTargetEvent.TargetReason.RANDOM_TARGET;
+
+                if (this instanceof EntityAIDefendVillage)
+                {
+                    reason = EntityTargetEvent.TargetReason.DEFEND_VILLAGE;
+                }
+                else if (this instanceof EntityAIHurtByTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY;
+                }
+                else if (this instanceof EntityAINearestAttackableTarget)
+                {
+                    if (p_75296_1_ instanceof EntityPlayer)
+                    {
+                        reason = EntityTargetEvent.TargetReason.CLOSEST_PLAYER;
+                    }
+                }
+                else if (this instanceof EntityAIOwnerHurtByTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER;
+                }
+                else if (this instanceof EntityAIOwnerHurtTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET;
+                }
+
+                org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(this.field_75299_d, p_75296_1_, reason);
+
+                if (event.isCancelled() || event.getTarget() == null)
+                {
+                    this.field_75299_d.func_70624_b(null);
+                    return false;
+                }
+                else if (p_75296_1_.getBukkitEntity() != event.getTarget())
+                {
+                    this.field_75299_d.func_70624_b((EntityLiving)((org.bukkit.craftbukkit.entity.CraftEntity) event.getTarget()).getHandle());
+                }
+
+                if (this.field_75299_d instanceof EntityCreature)
+                {
+                    ((EntityCreature) this.field_75299_d).field_70789_a = ((org.bukkit.craftbukkit.entity.CraftEntity) event.getTarget()).getHandle();
+                }
+
+                // CraftBukkit end
                 return true;
             }
         }
