--- ../net/minecraft/entity/ai/EntityAIAttackOnCollide.java
+++ ../net/minecraft/entity/ai/EntityAIAttackOnCollide.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.ai;
 
+import org.bukkit.event.entity.EntityTargetEvent; // CraftBukkit
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.pathfinding.PathEntity;
 import net.minecraft.util.MathHelper;
@@ -67,6 +68,10 @@
 
     public void func_75251_c()
     {
+        // CraftBukkit start
+        EntityTargetEvent.TargetReason reason = this.field_75442_c.func_70089_S() ? EntityTargetEvent.TargetReason.FORGOT_TARGET : EntityTargetEvent.TargetReason.TARGET_DIED;
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetEvent(field_75441_b, null, reason);
+        // CraftBukkit end
         this.field_75442_c = null;
         this.field_75441_b.func_70661_as().func_75499_g();
     }
