--- ../net/minecraft/entity/item/EntityEnderCrystal.java
+++ ../net/minecraft/entity/item/EntityEnderCrystal.java
@@ -6,7 +6,6 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
-
 public class EntityEnderCrystal extends Entity
 {
     public int field_70261_a = 0;
@@ -68,6 +67,13 @@
         {
             if (!this.field_70128_L && !this.field_70170_p.field_72995_K)
             {
+                // CraftBukkit start - All non-living entities need this
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_))
+                {
+                    return false;
+                }
+
+                // CraftBukkit end
                 this.field_70260_b = 0;
 
                 if (this.field_70260_b <= 0)
@@ -76,7 +82,7 @@
 
                     if (!this.field_70170_p.field_72995_K)
                     {
-                        this.field_70170_p.func_72876_a((Entity)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, 6.0F, true);
+                        this.field_70170_p.func_72876_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, 6.0F, true); // CraftBukkit - (Entity) null -> this
                     }
                 }
             }
