--- ../net/minecraft/entity/item/EntityMinecartContainer.java
+++ ../net/minecraft/entity/item/EntityMinecartContainer.java
@@ -1,5 +1,7 @@
 package net.minecraft.entity.item;
 
+// CraftBukkit start
+import java.util.List;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.IInventory;
@@ -9,10 +11,57 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.world.World;
 
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
+
 public abstract class EntityMinecartContainer extends EntityMinecart implements IInventory
 {
     private ItemStack[] field_94113_a = new ItemStack[36];
     private boolean field_94112_b = true;
+
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public ItemStack[] getContents()
+    {
+        return this.field_94113_a;
+    }
+
+    public void onOpen(CraftHumanEntity who)
+    {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who)
+    {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers()
+    {
+        return transaction;
+    }
+
+    public InventoryHolder getOwner()
+    {
+        org.bukkit.entity.Entity cart = getBukkitEntity();
+
+        if (cart instanceof InventoryHolder)
+        {
+            return (InventoryHolder) cart;
+        }
+
+        return null;
+    }
+
+    public void setMaxStackSize(int size)
+    {
+        maxStack = size;
+    }
+    // CraftBukkit end
 
     public EntityMinecartContainer(World p_i10051_1_)
     {
@@ -243,7 +292,7 @@
 
     protected void func_94101_h()
     {
-        int i = 15 - Container.func_94526_b(this);
+        int i = 15 - Container.func_94526_b((IInventory) this);
         float f = 0.98F + (float)i * 0.001F;
         this.field_70159_w *= (double)f;
         this.field_70181_x *= 0.0D;
