--- ../net/minecraft/entity/item/EntityItem.java
+++ ../net/minecraft/entity/item/EntityItem.java
@@ -9,10 +9,13 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.stats.AchievementList;
+import net.minecraft.stats.StatBase;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.MathHelper;
 import net.minecraft.util.StatCollector;
 import net.minecraft.world.World;
+
+import org.bukkit.event.player.PlayerPickupItemEvent; // CraftBukkit
 
 public class EntityItem extends Entity
 {
@@ -20,6 +23,7 @@
     public int field_70293_c;
     private int field_70291_e;
     public float field_70290_d;
+    private int lastTick = (int)(System.currentTimeMillis() / 50);  // CraftBukkit
 
     public EntityItem(World p_i8007_1_, double p_i8007_2_, double p_i8007_4_, double p_i8007_6_)
     {
@@ -39,6 +43,14 @@
     public EntityItem(World p_i3539_1_, double p_i3539_2_, double p_i3539_4_, double p_i3539_6_, ItemStack p_i3539_8_)
     {
         this(p_i3539_1_, p_i3539_2_, p_i3539_4_, p_i3539_6_);
+
+        // CraftBukkit start - Can't set null items in the datawatcher
+        if (p_i3539_8_ == null || p_i3539_8_.func_77973_b() == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
         this.func_92058_a(p_i3539_8_);
     }
 
@@ -65,12 +77,11 @@
     public void func_70071_h_()
     {
         super.func_70071_h_();
-
-        if (this.field_70293_c > 0)
-        {
-            --this.field_70293_c;
-        }
-
+        // CraftBukkit start
+        int currentTick = (int)(System.currentTimeMillis() / 50);
+        this.field_70293_c -= (currentTick - this.lastTick);
+        this.lastTick = currentTick;
+        // CraftBukkit end
         this.field_70169_q = this.field_70165_t;
         this.field_70167_r = this.field_70163_u;
         this.field_70166_s = this.field_70161_v;
@@ -121,6 +132,14 @@
 
         if (!this.field_70170_p.field_72995_K && this.field_70292_b >= 6000)
         {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled())
+            {
+                this.field_70292_b = 0;
+                return;
+            }
+
+            // CraftBukkit end
             this.func_70106_y();
         }
     }
@@ -194,7 +213,7 @@
 
     public boolean func_70072_I()
     {
-        return this.field_70170_p.func_72918_a(this.field_70121_D, Material.field_76244_g, this);
+        return this.field_70170_p.func_72918_a(this.field_70121_D, Material.field_76244_g, (Entity) this);
     }
 
     protected void func_70081_e(int p_70081_1_)
@@ -242,8 +261,27 @@
         this.field_70291_e = p_70037_1_.func_74765_d("Health") & 255;
         this.field_70292_b = p_70037_1_.func_74765_d("Age");
         NBTTagCompound nbttagcompound1 = p_70037_1_.func_74775_l("Item");
-        this.func_92058_a(ItemStack.func_77949_a(nbttagcompound1));
-
+
+        // CraftBukkit start
+        if (nbttagcompound1 != null)
+        {
+            ItemStack itemstack = ItemStack.func_77949_a(nbttagcompound1);
+
+            if (itemstack != null)
+            {
+                this.func_92058_a(itemstack);
+            }
+            else
+            {
+                this.func_70106_y();
+            }
+        }
+        else
+        {
+            this.func_70106_y();
+        }
+
+        // CraftBukkit end
         if (this.func_92059_d() == null)
         {
             this.func_70106_y();
@@ -256,27 +294,49 @@
         {
             ItemStack itemstack = this.func_92059_d();
             int i = itemstack.field_77994_a;
+            // CraftBukkit start
+            int canHold = p_70100_1_.field_71071_by.canHold(itemstack);
+            int remaining = itemstack.field_77994_a - canHold;
+
+            if (this.field_70293_c <= 0 && canHold > 0)
+            {
+                itemstack.field_77994_a = canHold;
+                PlayerPickupItemEvent event = new PlayerPickupItemEvent((org.bukkit.entity.Player) p_70100_1_.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                event.setCancelled(!p_70100_1_.field_82172_bs);
+                this.field_70170_p.getServer().getPluginManager().callEvent(event);
+                itemstack.field_77994_a = canHold + remaining;
+
+                if (event.isCancelled())
+                {
+                    return;
+                }
+
+                // Possibly < 0; fix here so we do not have to modify code below
+                this.field_70293_c = 0;
+            }
+
+            // CraftBukkit end
 
             if (this.field_70293_c == 0 && p_70100_1_.field_71071_by.func_70441_a(itemstack))
             {
                 if (itemstack.field_77993_c == Block.field_71951_J.field_71990_ca)
                 {
-                    p_70100_1_.func_71029_a(AchievementList.field_76005_g);
+                    p_70100_1_.func_71029_a((StatBase) AchievementList.field_76005_g);
                 }
 
                 if (itemstack.field_77993_c == Item.field_77770_aF.field_77779_bT)
                 {
-                    p_70100_1_.func_71029_a(AchievementList.field_76022_t);
+                    p_70100_1_.func_71029_a((StatBase) AchievementList.field_76022_t);
                 }
 
                 if (itemstack.field_77993_c == Item.field_77702_n.field_77779_bT)
                 {
-                    p_70100_1_.func_71029_a(AchievementList.field_76019_w);
+                    p_70100_1_.func_71029_a((StatBase) AchievementList.field_76019_w);
                 }
 
                 if (itemstack.field_77993_c == Item.field_77731_bo.field_77779_bT)
                 {
-                    p_70100_1_.func_71029_a(AchievementList.field_76027_z);
+                    p_70100_1_.func_71029_a((StatBase) AchievementList.field_76027_z);
                 }
 
                 this.func_85030_a("random.pop", 0.2F, ((this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.7F + 1.0F) * 2.0F);
