--- ../net/minecraft/entity/monster/EntityMagmaCube.java
+++ ../net/minecraft/entity/monster/EntityMagmaCube.java
@@ -2,7 +2,6 @@
 
 import net.minecraft.item.Item;
 import net.minecraft.world.World;
-
 public class EntityMagmaCube extends EntitySlime
 {
     public EntityMagmaCube(World p_i3551_1_)
@@ -45,6 +44,8 @@
 
     protected void func_70628_a(boolean p_70628_1_, int p_70628_2_)
     {
+        // CraftBukkit start - whole method
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.func_70633_aT();
 
         if (j > 0 && this.func_70809_q() > 1)
@@ -56,11 +57,14 @@
                 k += this.field_70146_Z.nextInt(p_70628_2_ + 1);
             }
 
-            for (int l = 0; l < k; ++l)
+            if (k > 0)
             {
-                this.func_70025_b(j, 1);
+                loot.add(new org.bukkit.inventory.ItemStack(j, k));
             }
         }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     public boolean func_70027_ad()
