--- ../net/minecraft/block/BlockRedstoneOre.java
+++ ../net/minecraft/block/BlockRedstoneOre.java
@@ -7,6 +7,8 @@
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+
+import org.bukkit.event.entity.EntityInteractEvent; // CraftBukkit
 
 public class BlockRedstoneOre extends Block
 {
@@ -37,8 +39,30 @@
 
     public void func_71891_b(World p_71891_1_, int p_71891_2_, int p_71891_3_, int p_71891_4_, Entity p_71891_5_)
     {
-        this.func_72176_l(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_);
-        super.func_71891_b(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_, p_71891_5_);
+        // CraftBukkit start
+        if (p_71891_5_ instanceof EntityPlayer)
+        {
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((EntityPlayer) p_71891_5_, org.bukkit.event.block.Action.PHYSICAL, p_71891_2_, p_71891_3_, p_71891_4_, -1, null);
+
+            if (!event.isCancelled())
+            {
+                this.func_72176_l(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_);
+                super.func_71891_b(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_, p_71891_5_);
+            }
+        }
+        else
+        {
+            EntityInteractEvent event = new EntityInteractEvent(p_71891_5_.getBukkitEntity(), p_71891_1_.getWorld().getBlockAt(p_71891_2_, p_71891_3_, p_71891_4_));
+            p_71891_1_.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled())
+            {
+                this.func_72176_l(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_);
+                super.func_71891_b(p_71891_1_, p_71891_2_, p_71891_3_, p_71891_4_, p_71891_5_);
+            }
+        }
+
+        // CraftBukkit end
     }
 
     public boolean func_71903_a(World p_71903_1_, int p_71903_2_, int p_71903_3_, int p_71903_4_, EntityPlayer p_71903_5_, int p_71903_6_, float p_71903_7_, float p_71903_8_, float p_71903_9_)
@@ -83,12 +107,24 @@
     public void func_71914_a(World p_71914_1_, int p_71914_2_, int p_71914_3_, int p_71914_4_, int p_71914_5_, float p_71914_6_, int p_71914_7_)
     {
         super.func_71914_a(p_71914_1_, p_71914_2_, p_71914_3_, p_71914_4_, p_71914_5_, p_71914_6_, p_71914_7_);
+        /* CraftBukkit start - delegate to getExpDrop
+        if (this.getDropType(l, world.random, i1) != this.id) {
+            int j1 = 1 + world.random.nextInt(5);
 
-        if (this.func_71885_a(p_71914_5_, p_71914_1_.field_73012_v, p_71914_7_) != this.field_71990_ca)
+            this.j(world, i, j, k, j1);
+        } */
+    }
+
+    public int getExpDrop(World world, int l, int i1)
+    {
+        if (this.func_71885_a(l, world.field_73012_v, i1) != this.field_71990_ca)
         {
-            int j1 = 1 + p_71914_1_.field_73012_v.nextInt(5);
-            this.func_71923_g(p_71914_1_, p_71914_2_, p_71914_3_, p_71914_4_, j1);
+            int j1 = 1 + world.field_73012_v.nextInt(5);
+            return j1;
         }
+
+        return 0;
+        // CraftBukkit end
     }
 
     private void func_72177_n(World p_72177_1_, int p_72177_2_, int p_72177_3_, int p_72177_4_)
