--- ../net/minecraft/item/crafting/IRecipe.java
+++ ../net/minecraft/item/crafting/IRecipe.java
@@ -3,7 +3,6 @@
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
-
 public interface IRecipe
 {
     boolean func_77569_a(InventoryCrafting inventorycrafting, World world);
@@ -13,4 +12,6 @@
     int func_77570_a();
 
     ItemStack func_77571_b();
+
+    org.bukkit.inventory.Recipe toBukkitRecipe(); // CraftBukkit
 }
