--- ../net/minecraft/item/BehaviorDispenseArmor.java
+++ ../net/minecraft/item/BehaviorDispenseArmor.java
@@ -2,16 +2,26 @@
 
 import java.util.List;
 import net.minecraft.block.BlockDispenser;
+import net.minecraft.command.IEntitySelector;
 import net.minecraft.dispenser.BehaviorDefaultDispenseItem;
+import net.minecraft.dispenser.IBehaviorDispenseItem;
 import net.minecraft.dispenser.IBlockSource;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntitySelectorArmoredMob;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.util.EnumFacing;
+import net.minecraft.world.World;
+
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
 
 final class BehaviorDispenseArmor extends BehaviorDefaultDispenseItem
 {
+    BehaviorDispenseArmor() {}
+
     protected ItemStack func_82487_b(IBlockSource p_82487_1_, ItemStack p_82487_2_)
     {
         EnumFacing enumfacing = BlockDispenser.func_100009_j_(p_82487_1_.func_82620_h());
@@ -19,18 +29,50 @@
         int j = p_82487_1_.func_82622_e() + enumfacing.func_96559_d();
         int k = p_82487_1_.func_82621_f() + enumfacing.func_82599_e();
         AxisAlignedBB axisalignedbb = AxisAlignedBB.func_72332_a().func_72299_a((double)i, (double)j, (double)k, (double)(i + 1), (double)(j + 1), (double)(k + 1));
-        List list = p_82487_1_.func_82618_k().func_82733_a(EntityLiving.class, axisalignedbb, new EntitySelectorArmoredMob(p_82487_2_));
+        List list = p_82487_1_.func_82618_k().func_82733_a(EntityLiving.class, axisalignedbb, (IEntitySelector)(new EntitySelectorArmoredMob(p_82487_2_)));
 
         if (list.size() > 0)
         {
             EntityLiving entityliving = (EntityLiving)list.get(0);
             int l = entityliving instanceof EntityPlayer ? 1 : 0;
             int i1 = EntityLiving.func_82159_b(p_82487_2_);
-            ItemStack itemstack1 = p_82487_2_.func_77946_l();
+            // CraftBukkit start
+            ItemStack itemstack1 = p_82487_2_.func_77979_a(1);
+            World world = p_82487_1_.func_82618_k();
+            org.bukkit.block.Block block = world.getWorld().getBlockAt(p_82487_1_.func_82623_d(), p_82487_1_.func_82622_e(), p_82487_1_.func_82621_f());
+            CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+            BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+
+            if (!BlockDispenser.eventFired)
+            {
+                world.getServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled())
+            {
+                p_82487_2_.field_77994_a++;
+                return p_82487_2_;
+            }
+
+            if (!event.getItem().equals(craftItem))
+            {
+                p_82487_2_.field_77994_a++;
+                // Chain to handler for new item
+                ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+                IBehaviorDispenseItem ibehaviordispenseitem = (IBehaviorDispenseItem) BlockDispenser.field_82527_a.func_82594_a(eventStack.func_77973_b());
+
+                if (ibehaviordispenseitem != IBehaviorDispenseItem.field_82483_a && ibehaviordispenseitem != this)
+                {
+                    ibehaviordispenseitem.func_82482_a(p_82487_1_, eventStack);
+                    return p_82487_2_;
+                }
+            }
+
+            // CraftBukkit end
             itemstack1.field_77994_a = 1;
             entityliving.func_70062_b(i1 - l, itemstack1);
             entityliving.func_96120_a(i1, 2.0F);
-            --p_82487_2_.field_77994_a;
+            // --itemstack.count; // CraftBukkit - handled above
             return p_82487_2_;
         }
         else
