--- ../net/minecraft/item/ItemSkull.java
+++ ../net/minecraft/item/ItemSkull.java
@@ -9,7 +9,6 @@
 import net.minecraft.util.MathHelper;
 import net.minecraft.util.StatCollector;
 import net.minecraft.world.World;
-
 public class ItemSkull extends Item
 {
     private static final String[] field_82807_a = new String[] {"skeleton", "wither", "zombie", "char", "creeper"};
@@ -25,6 +24,8 @@
 
     public boolean func_77648_a(ItemStack p_77648_1_, EntityPlayer p_77648_2_, World p_77648_3_, int p_77648_4_, int p_77648_5_, int p_77648_6_, int p_77648_7_, float p_77648_8_, float p_77648_9_, float p_77648_10_)
     {
+        final int clickedX = p_77648_4_, clickedY = p_77648_5_, clickedZ = p_77648_6_; // CraftBukkit
+
         if (p_77648_7_ == 0)
         {
             return false;
@@ -70,7 +71,15 @@
             }
             else
             {
-                p_77648_3_.func_72832_d(p_77648_4_, p_77648_5_, p_77648_6_, Block.field_82512_cj.field_71990_ca, p_77648_7_, 2);
+                // CraftBukkit start - handle in ItemBlock
+                // world.setTypeIdAndData(i, j, k, Block.SKULL.id, l, 2);
+                if (!ItemBlock.processBlockPlace(p_77648_3_, p_77648_2_, null, p_77648_4_, p_77648_5_, p_77648_6_, Block.field_82512_cj.field_71990_ca, p_77648_7_, clickedX, clickedY, clickedZ))
+                {
+                    return false;
+                }
+
+                p_77648_7_ = p_77648_3_.func_72805_g(p_77648_4_, p_77648_5_, p_77648_6_);
+                // CraftBukkit end
                 int i1 = 0;
 
                 if (p_77648_7_ == 1)
