--- ../net/minecraft/item/ItemLilyPad.java
+++ ../net/minecraft/item/ItemLilyPad.java
@@ -6,7 +6,6 @@
 import net.minecraft.util.EnumMovingObjectType;
 import net.minecraft.util.MovingObjectPosition;
 import net.minecraft.world.World;
-
 public class ItemLilyPad extends ItemColored
 {
     public ItemLilyPad(int p_i3694_1_)
@@ -29,6 +28,7 @@
                 int i = movingobjectposition.field_72311_b;
                 int j = movingobjectposition.field_72312_c;
                 int k = movingobjectposition.field_72309_d;
+                final int clickedX = i, clickedY = j, clickedZ = k; // CraftBukkit
 
                 if (!p_77659_2_.func_72962_a(p_77659_3_, i, j, k))
                 {
@@ -42,7 +42,14 @@
 
                 if (p_77659_2_.func_72803_f(i, j, k) == Material.field_76244_g && p_77659_2_.func_72805_g(i, j, k) == 0 && p_77659_2_.func_72799_c(i, j + 1, k))
                 {
-                    p_77659_2_.func_94575_c(i, j + 1, k, Block.field_71991_bz.field_71990_ca);
+                    // CraftBukkit start - waterlily
+                    // world.setTypeIdUpdate(i, j + 1, k, Block.WATER_LILY.id);
+                    if (!processBlockPlace(p_77659_2_, p_77659_3_, null, i, j + 1, k, Block.field_71991_bz.field_71990_ca, 0, clickedX, clickedY, clickedZ))
+                    {
+                        return p_77659_1_;
+                    }
+
+                    // CraftBukkit end
 
                     if (!p_77659_3_.field_71075_bZ.field_75098_d)
                     {
